MediaRecorder.html

```diff
@@ -92,22 +92,30 @@
-      <li>Then fire a <code>dataavailable</code> event containing the Blob of data.</li>
-      <li>Finally, fire a <code>stop</code> event.</li>
-     </ol>
+    <dd>When a <code>MediaRecorder</code> object’s <code>start()</code> method is invoked, the UA <em title="must" class="rfc2119">must</em> run the following steps:
+      <ol>
+        <li>Let <var>timeslice</var> be the method's first argument, if provided, or <code>undefined</code>.</li>
+        <li>If <code>state</code> is not "inactive", throw an <code>InvalidStateError</code> DOMException and abort these steps.</li>
+        <li>If the stream's isolation properties disallow access from this MediaRecorder, throw a <code>SecurityError</code> DOMException and abort these steps.</li>
+        <li>Set <code>state</code> to "recording", and run the following steps in parallel:</li>
+        <ol>
+           <li>Once media becomes available from one or more of the stream's tracks, start gathering the data into a Blob (see [[!FILE-API]]), and queue a task, using the DOM manipulation task source, to fire a <code>start</code> event. </li>
+           <li>If at any point the stream's isolation properties change so that MediaRecorder is no longer allowed access to it, the UA <em title="must" class="rfc2119">must</em> immediately stop gathering data, discard any data that it has gathered, and queue a task, using the DOM manipulation task source, that <a href="#track-ended">fires an error event</a> with the name <code>SecurityError</code>, immediately followed by a <code>stop</code> event.
+           <li>If the UA at any point is unable to continue gathering data for reasons other than isolation properties, it <em title="must" class="rfc2119">must</em> stop gathering data, let <var>blob</var> be the Blob of collected data, and queue a task, using the DOM manipulation task source, that <a href="#track-ended">fires an error event</a> with the name <code>UnknownError</code>, followed by a <code>dataavailable</code> event containing <var>blob</var>, follwed by a <code>stop</code> event.</li>
+           <li>If <var>timeslice</var> is not <code>undefined</code>, then once a minimum of <var>timeslice</var> milliseconds of data have been collected, or some minimum time slice imposed by the UA, whichever is greater, let <var>blob</var> be the Blob of collected data, start gathering a new Blob of data, and queue a task, using the DOM manipulation task source, that fires a <code>dataavailable</code> event containing <var>blob</var>.</li>
+           <li>If all recorded tracks become <a href="/mediacapture-main/getusermedia.html#track-ended">ended</a>, then stop gathering data, let <var>blob</var> be the Blob of collected data, and queue a task, using the DOM manipulation task source, that runs the following steps:</li>
```

Comment:
This looks good to me, using the DOM manipulation task source is probably the right choice as long as it's never possible that the events which have already been queued should be removed from the queue. But... I'm not certain that's the case. If one calls stop() and start() in quick succession with unfortunate timing, one could get the dataavailable and stop events from before the stop and have no way of knowing what they mean.

The way that media elements deal with a similar situation is to have a task source per instance, the "media element task source". It's kind of annoying, but if you have that kind of lifecycle then I don't see another way to make the API robust.